karoo_gp/base_class.py

```diff
@@ -211,68 +194,141 @@ def error(self, msg, display={'i', 'g', 'm', 'db'}):
-                self.log(f'\n Evolve a population of Trees for Generation {gen + 1} ...')
+
+                # Evolve the next generation
+                self.log(f'Evolve a population for Generation {gen + 1} ...')
                 self.population = self.population.evolve(
-                    self.swim, self.tree_depth_min, self.functions,
-                    self.terminals, self.evolve_repro, self.evolve_point,
-                    self.evolve_branch, self.evolve_cross, self.tree_pop_max,
-                    self.tourn_size, self.precision, self.fitness_type,
-                    self.tree_depth_max, self.data_train, self.kernel,
-                    self.data_train_rows, self.tf_device_log, self.class_labels,
-                    self.tf_device, self.savefile, self.rng, self.log,
-                    self.pause, self.error, self.fx_data_tree_write,
+                    self.tree_pop_max,
+                    self.functions,
+                    self.terminals,
+                    self.swim,
+                    self.tree_depth_min,
+                    self.tree_depth_max,
+                    self.tourn_size,
+                    self.evolve_repro,
+                    self.evolve_point,
+                    self.evolve_branch,
+                    self.evolve_cross,
                 )
-                self.log('\n Copy gp.population_b to gp.population_a\n')
 
-            if self.mode == 's':
-                # (s)erver mode - termination with completiont of prescribed run
+                # Evaluate new generation
+                self.population.evaluate(X_train, y_train, X_train_hash)
+
+                # Add best score to history
+                for k, v in self.score().items():
+                    self.history[k].append(v)
+
+            if self.mode == 's':  # (s)erver mode: terminate after run
                 menu = 0
-            else:
-                # (d)esktop mode - user is given an option to quit, review,
-                # and/or modify parameters; 'add' generations continues the run
-                self.log('\n\t\033[32m Enter \033[1m?\033[0;0m\033[32m to review '
-                         'your options or \033[1mq\033[0;0m\033[32muit\033[0;0m')
+            else:                 # (d)esktop mode: pause after run
+                self.log('Enter ? to review your options or q to quit')
                 menu = self.pause()
 
+    def predict(self, X, trees=None, X_hash=None):
+        """Return predicted y values for X
+
+        * Use fittest tree by default (for sklearn `estimator.predict`)
+        * Also accepts a tree or list of trees (for population.evaluate)
+        * Output type (single or a list) matches input type
+        * If X_hash is provided, return zeros for trees with cached scores
+        # Encode and round, if necessary
+        """
+        trees = self.population.fittest() if trees is None else trees
+        trees = [trees] if type(trees) != list else trees
```

Comment:
See what happens when you have functions that return different types? :stuck_out_tongue:
"Special cases aren't special enough to break the rules."

Reply:
I've redone this now to look like:
```
def predict(self, X):
    """Return predicted y values for X using the fittest tree

    * Pass the fittest tree to batch_predict as a list (of 1)
    * Return the first result
    """
    fittest_tree = self.population.fittest()
    return self.batch_predict(X, [fittest_tree])[0]

def batch_predict(self, X, trees, X_hash=None):
    """Return predicted values for y given X for a list of trees

    * If encoder (e.g. MultiClassifier), decode predictions
    * If precision is specified, round predictions to precision
    """
    self.engine._type_check(trees)
    y = self.engine.predict(trees, X, X_hash)
    if self.encoder is not None:
        y = self.encoder.decode(y)
    if self.precision is not None:
        y = np.round(y, self.precision)
    return y
```
Will do something similar with score to make the inputs consistent.